ipyparallel/client/magics.py

```diff
@@ -364,49 +364,52 @@ def parallel_execute(
-                        result.wait_interactive()
-                    result.wait_for_output()
-                    try:
-                        result.get()
-                    except error.CompositeError as e:
-                        if stream_output:
-                            # already streamed, show an abbreviated result
-                            raise error.AlreadyDisplayedError(e) from None
+            try:
+                if progress_after is None:
+                    progress_after = self.progress_after_seconds
+
+                cm = result.stream_output() if stream_output else nullcontext()
+                with cm:
+                    finished_waiting = False
+                    if progress_after > 0:
+                        # finite progress-after timeout
+                        # wait for 'quick' results before showing progress
+                        tic = time.perf_counter()
+                        deadline = tic + progress_after
+                        try:
+                            result.get(timeout=progress_after)
+                            remaining = max(deadline - time.perf_counter(), 0)
+                            result.wait_for_output(timeout=remaining)
+                        except TimeoutError:
+                            pass
+                        except error.CompositeError as e:
+                            if stream_output:
+                                # already streamed, show an abbreviated result
+                                raise error.AlreadyDisplayedError(e) from None
+                            else:
+                                raise
                         else:
-                            raise
-            # Skip redisplay if streaming output
-            if not stream_output:
-                result.display_outputs(groupby)
+                            finished_waiting = True
+
+                    if not finished_waiting:
+                        if progress_after >= 0:
+                            # not an immediate result, start interactive progress
+                            result.wait_interactive()
+                        result.wait_for_output()
+                        try:
+                            result.get()
+                        except error.CompositeError as e:
+                            if stream_output:
+                                # already streamed, show an abbreviated result
+                                raise error.AlreadyDisplayedError(e) from None
+                            else:
+                                raise
+                # Skip redisplay if streaming output
+                if not stream_output:
+                    result.display_outputs(groupby)
+            except KeyboardInterrupt:
+                print("Received Keyboard Interrupt. Sending signal 9 to engines...")
+                self.view.client.send_signal(9, targets=targets, block=True)
```

Comment:
SIGKILL is a very harsh consequence for impatience. I definitely don't want that to be unconditional. Usually you'd want signal 2 (SIGINT). I realize for MPI that SIGINT isn't always enough, but SIGKILL should be a last resort. Even then, I'm not sure we want it to be unconditional. Maybe a px option, e.g. `--signal-on-interrupt=2|INT|KILL`?

Reply:
Makes sense. I'll add that option. Should the default be to do nothing (like now), or SIGINT?

Reply:
Added the option.